**Day 1** - Initial Setup


In [1]:
# Day 1 - Initial Setup
import pandas as pd
train_df = pd.read_csv("cohort_train.csv")

train_df.head()



,learner_id,signup_date,age,city_tier,education,hours_per_week_pledged,weekly_logins_avg,assignments_submitted,forum_posts,video_watch_pct,has_mentor,device,referral_source,payment_plan,days_since_last_login,certificate_issued,final_score,notes,completed_course
0,L-01146,2026-01-22,24.0,Tier 3,OTHER,8.3,4.20,3,6,89.0,Y,MOBILE,instagram,scholarship,1,Y,78.6,NaN,1.0
1,L-00695,2026-06-02,15.0,Tier 1,Bachelors,1.0,2.54,0,4,31.2,no,desktop,instagram,paid,33,no,NaN,NaN,0.0
2,L-00530,20-05-2026,24.0,tier-2,Diploma,10.3,0.50,0,3,10.7,NaN,laptop,instagram,scholarship,23,No,NaN,follow up next intake,0.0
3,L-02228,2026-01-19,23.0,1,B.Tech,1.6,3.01,0,0,40.2,no,mobile,school_partner,free,33,FALSE,NaN,duplicate? check,0.0
4,L-00914,16/02/2026,21.0,2,High School,9.8,4.36,3,1,36.7,no,Desktop,instagram,paid,19,TRUE,67.1,follow up next intake,1.0


In [2]:
train_df.shape

(2484, 19)

In [3]:
train_df = pd.read_csv("cohort_train.csv")
train_df.columns

Index(['learner_id', 'signup_date', 'age', 'city_tier', 'education',
       'hours_per_week_pledged', 'weekly_logins_avg', 'assignments_submitted',
       'forum_posts', 'video_watch_pct', 'has_mentor', 'device',
       'referral_source', 'payment_plan', 'days_since_last_login',
       'certificate_issued', 'final_score', 'notes', 'completed_course'],
      dtype='object')

**Day 2** - Data Exploration


In [4]:
# Day 2 - Data Exploration
train_df.isnull().sum()

,0
learner_id,0
signup_date,0
age,152
city_tier,53
education,90
hours_per_week_pledged,108
weekly_logins_avg,96
assignments_submitted,0
forum_posts,0
video_watch_pct,147


In [5]:
missing = train_df.isnull().sum()

missing_percent = (missing / len(train_df)) * 100

missing_df = pd.DataFrame({
    "Missing Values": missing,
    "Percentage": missing_percent
})

missing_df.sort_values(by="Missing Values", ascending=False)

,Missing Values,Percentage
final_score,1322,53.220612
notes,1244,50.080515
age,152,6.119163
video_watch_pct,147,5.917874
hours_per_week_pledged,108,4.347826
weekly_logins_avg,96,3.864734
education,90,3.623188
has_mentor,64,2.576490
city_tier,53,2.133655
completed_course,9,0.362319


In [6]:
train_df.duplicated().sum()

np.int64(83)

In [7]:
for column in train_df.columns:
    print(column)

learner_id
signup_date
age
city_tier
education
hours_per_week_pledged
weekly_logins_avg
assignments_submitted
forum_posts
video_watch_pct
has_mentor
device
referral_source
payment_plan
days_since_last_login
certificate_issued
final_score
notes
completed_course


In [8]:
train_df.select_dtypes(include="object").columns

Index(['learner_id', 'signup_date', 'city_tier', 'education', 'has_mentor',
       'device', 'referral_source', 'payment_plan', 'certificate_issued',
       'notes'],
      dtype='object')

In [9]:
for column in train_df.select_dtypes(include="object").columns:
    print("=" * 50)
    print(column)
    print(train_df[column].value_counts(dropna=False))

learner_id
learner_id
L-01799    2
L-02256    2
L-02454    2
L-00414    2
L-01425    2
          ..
L-02488    1
L-02423    1
L-00398    1
L-02542    1
L-00467    1
Name: count, Length: 2400, dtype: int64
signup_date
signup_date
2026-01-21    19
2026-01-18    19
2026-06-14    18
2026-01-04    18
2026-06-08    18
              ..
10/02/2026     1
14/02/2026     1
21-04-2026     1
07/06/2026     1
10/05/2026     1
Name: count, Length: 462, dtype: int64
city_tier
city_tier
2         793
1         663
3         498
T2         71
Tier 2     68
tier-2     65
Tier 1     57
tier-1     57
NaN        53
T1         46
tier-3     40
Tier 3     38
T3         35
Name: count, dtype: int64
education
education
Bachelors          629
High School        587
Diploma            396
Masters            173
Other              107
NaN                 90
B.Tech              52
highschool          50
DIPLOMA             50
HS                  44
Bachelor's          40
bachelors           40
diploma             3

In [10]:
train_df.describe().T

,count,mean,std,min,25%,50%,75%,max
age,2332.0,22.714408,20.278475,-3.0,18.000,21.000,24.0000,300.0
hours_per_week_pledged,2376.0,8.437626,18.435356,1.0,4.800,7.100,9.5000,326.8
weekly_logins_avg,2388.0,3.075096,1.502140,0.0,2.070,3.125,4.0825,7.0
assignments_submitted,2484.0,3.655797,2.493706,0.0,2.000,4.000,5.0000,10.0
forum_posts,2484.0,3.247585,2.505309,0.0,1.000,3.000,5.0000,14.0
video_watch_pct,2337.0,46.617758,25.264902,0.0,28.400,45.400,63.9000,115.3
days_since_last_login,2484.0,19.288647,11.864180,0.0,10.000,18.000,27.0000,70.0
final_score,1162.0,70.232530,16.816240,13.3,62.025,74.250,81.7750,100.0
completed_course,2475.0,0.369293,0.482711,0.0,0.000,0.000,1.0000,1.0


In [11]:
train_df[["age",
          "hours_per_week_pledged",
          "weekly_logins_avg",
          "video_watch_pct"]].describe()

,age,hours_per_week_pledged,weekly_logins_avg,video_watch_pct
count,2332.000000,2376.000000,2388.000000,2337.000000
mean,22.714408,8.437626,3.075096,46.617758
std,20.278475,18.435356,1.502140,25.264902
min,-3.000000,1.000000,0.000000,0.000000
25%,18.000000,4.800000,2.070000,28.400000
50%,21.000000,7.100000,3.125000,45.400000
75%,24.000000,9.500000,4.082500,63.900000
max,300.000000,326.800000,7.000000,115.300000


In [12]:
train_df.describe().T

,count,mean,std,min,25%,50%,75%,max
age,2332.0,22.714408,20.278475,-3.0,18.000,21.000,24.0000,300.0
hours_per_week_pledged,2376.0,8.437626,18.435356,1.0,4.800,7.100,9.5000,326.8
weekly_logins_avg,2388.0,3.075096,1.502140,0.0,2.070,3.125,4.0825,7.0
assignments_submitted,2484.0,3.655797,2.493706,0.0,2.000,4.000,5.0000,10.0
forum_posts,2484.0,3.247585,2.505309,0.0,1.000,3.000,5.0000,14.0
video_watch_pct,2337.0,46.617758,25.264902,0.0,28.400,45.400,63.9000,115.3
days_since_last_login,2484.0,19.288647,11.864180,0.0,10.000,18.000,27.0000,70.0
final_score,1162.0,70.232530,16.816240,13.3,62.025,74.250,81.7750,100.0
completed_course,2475.0,0.369293,0.482711,0.0,0.000,0.000,1.0000,1.0


In [13]:
train_df[train_df["age"] < 0][["learner_id", "age"]]

,learner_id,age
132,L-02530,-3.0
895,L-00503,-3.0
1090,L-01264,-3.0
1243,L-00261,-3.0
1469,L-02080,-3.0
1682,L-01445,-3.0


In [14]:
train_df[train_df["age"] > 100][["learner_id", "age"]]

,learner_id,age
30,L-02703,300.0
281,L-01185,199.0
481,L-02448,300.0
531,L-02461,300.0
622,L-00842,300.0
713,L-01071,199.0
841,L-01721,300.0
889,L-00090,300.0
902,L-01128,300.0
1254,L-01682,199.0


In [15]:
train_df[train_df["hours_per_week_pledged"] > 168][
    ["learner_id", "hours_per_week_pledged"]
]

,learner_id,hours_per_week_pledged
48,L-01420,318.3
178,L-02033,319.0
351,L-02360,233.8
663,L-00722,325.3
1351,L-00249,300.9
1441,L-02268,326.8
1553,L-02345,220.3
1624,L-01157,272.8
1854,L-02345,220.3


In [16]:
train_df[train_df["hours_per_week_pledged"] <= 0][
    ["learner_id", "hours_per_week_pledged"]
]

,learner_id,hours_per_week_pledged


In [17]:
train_df[train_df["days_since_last_login"] > 21][
    ["learner_id", "days_since_last_login"]
]

,learner_id,days_since_last_login
1,L-00695,33
2,L-00530,23
3,L-02228,33
7,L-02644,25
8,L-02550,24
...,...,...
2472,L-00678,29
2473,L-00973,31
2477,L-01332,32
2480,L-01303,32


In [18]:
train_df[train_df["video_watch_pct"] > 100][
    ["learner_id", "video_watch_pct"]
]

,learner_id,video_watch_pct
15,L-02774,101.0
37,L-01194,102.6
107,L-01358,112.9
127,L-00025,104.2
155,L-00162,105.9
293,L-01947,107.9
350,L-01573,103.2
402,L-01230,113.1
459,L-00480,107.6
565,L-02712,110.1


In [19]:
train_df["learner_id"].duplicated().sum()

np.int64(84)

In [20]:
train_df["learner_id"].value_counts().head(20)

,count
learner_id,
L-01799,2
L-02256,2
L-02454,2
L-00414,2
L-01425,2
L-01320,2
L-02286,2
L-01352,2
L-00056,2


In [21]:
duplicate_ids = train_df[train_df["learner_id"].duplicated(keep=False)].sort_values("learner_id")

duplicate_ids[["learner_id", "age", "education", "completed_course"]].head(30)

,learner_id,age,education,completed_course
1417,L-00002,22.0,Diploma,0.0
1245,L-00002,22.0,Diploma,0.0
856,L-00056,22.0,Other,0.0
1570,L-00056,22.0,Other,0.0
2481,L-00082,28.0,Bachelors,1.0
651,L-00082,28.0,Bachelors,1.0
263,L-00141,25.0,Bachelors,1.0
156,L-00141,25.0,Bachelors,1.0
167,L-00229,15.0,NaN,1.0
172,L-00229,15.0,NaN,1.0


In [22]:
train_df.duplicated().sum()

np.int64(83)

In [23]:
duplicate_ids.groupby("learner_id").nunique().sort_values(
    by="learner_id"
).head(50)

,signup_date,age,city_tier,education,hours_per_week_pledged,weekly_logins_avg,assignments_submitted,forum_posts,video_watch_pct,has_mentor,device,referral_source,payment_plan,days_since_last_login,certificate_issued,final_score,notes,completed_course
learner_id,,,,,,,,,,,,,,,,,,
L-00002,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,0,1
L-00056,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,0,1
L-00082,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1
L-00141,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1
L-00229,1,1,1,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1
L-00268,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,0,1
L-00406,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,0,1
L-00414,1,1,1,0,1,1,1,1,1,1,1,1,1,1,1,0,1,1
L-00453,1,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,1


In [24]:
conflicting_ids = []

for learner_id, group in duplicate_ids.groupby("learner_id"):
    if group.drop(columns=["learner_id"]).nunique(dropna=False).max() > 1:
        conflicting_ids.append(learner_id)

conflicting_ids

['L-01467']

In [25]:
train_df[train_df["learner_id"].isin(conflicting_ids)].T

,551,939
learner_id,L-01467,L-01467
signup_date,27-05-2026,27-05-2026
age,28.0,28.0
city_tier,Tier 2,Tier 2
education,Bachelors,Bachelors
hours_per_week_pledged,13.6,13.6
weekly_logins_avg,2.89,2.89
assignments_submitted,6,6
forum_posts,1,1
video_watch_pct,63.5,63.5


In [26]:
train_df["education"].value_counts(dropna=False)

,count
education,
Bachelors,629
High School,587
Diploma,396
Masters,173
Other,107
NaN,90
B.Tech,52
highschool,50
DIPLOMA,50


In [27]:
train_df["city_tier"].value_counts(dropna=False)

,count
city_tier,
2,793
1,663
3,498
T2,71
Tier 2,68
tier-2,65
Tier 1,57
tier-1,57
NaN,53


In [28]:
train_df["has_mentor"].value_counts(dropna=False)

,count
has_mentor,
0,310
No,299
FALSE,295
no,293
N,290
1,199
TRUE,191
Yes,187
Y,184


In [29]:
train_df["device"].value_counts(dropna=False)

,count
device,
mobile,1017
laptop,599
desktop,232
tablet,119
Mobile,101
phone,88
MOBILE,86
notebook,74
Laptop,68


In [30]:
train_df["referral_source"].value_counts(dropna=False)

,count
referral_source,
instagram,687
school_partner,554
friend,475
youtube,367
ngo_outreach,265
other,136


In [31]:
train_df["payment_plan"].value_counts(dropna=False)

,count
payment_plan,
free,1130
scholarship,811
paid,543


In [32]:
train_df["signup_date"].head(20)

,signup_date
0,2026-01-22
1,2026-06-02
2,20-05-2026
3,2026-01-19
4,16/02/2026
5,2026-06-03
6,2026-04-09
7,11-04-2026
8,2026-04-14
9,09/01/2026


In [33]:
train_df["signup_date"].dtype

dtype('O')

In [34]:
train_df["device"].value_counts(dropna=False)

,count
device,
mobile,1017
laptop,599
desktop,232
tablet,119
Mobile,101
phone,88
MOBILE,86
notebook,74
Laptop,68


In [35]:
train_df["city_tier"].value_counts(dropna=False)

,count
city_tier,
2,793
1,663
3,498
T2,71
Tier 2,68
tier-2,65
Tier 1,57
tier-1,57
NaN,53


In [36]:
train_df["signup_date"].min(), train_df["signup_date"].max()

('01-01-2026', '28/06/2026')

# Day 2 - Data Exploration Findings

## Data Quality Issues Discovered

During exploration of the training dataset, several data quality issues were identified.

### 1. Missing Values
Several useful learner features contain missing values. The highest missingness among potentially useful features is in `age` (152 rows, 6.12%), followed by `video_watch_pct` (147 rows, 5.92%), `hours_per_week_pledged` (108 rows, 4.35%), `weekly_logins_avg` (96 rows, 3.86%), `education` (90 rows, 3.62%), `has_mentor` (64 rows, 2.58%), and `city_tier` (53 rows, 2.13%).

There are also 9 rows where `completed_course`, the target variable, is missing. These rows cannot provide a known outcome for supervised training.

### 2. Duplicate Records
84 rows contain learner IDs that appear more than once. Of these, 83 are exact duplicate rows. One learner, `L-01467`, appears twice with identical feature information, but one record has a missing `completed_course` value while the other has `completed_course = 1`.

The duplicate records are consistent with the export issue described in the Data Dictionary.

### 3. Invalid Age Values
The `age` column contains impossible values. Six learners have an age of `-3`, while 16 learners have ages of `199` or `300`. These values are inconsistent with realistic learner ages and will need to be handled during cleaning.

### 4. Impossible Weekly Hours
The `hours_per_week_pledged` column contains 9 values greater than 168 hours per week. Since a week contains only 168 hours, these values are physically impossible and indicate invalid input data.

### 5. Unexpected Login-Day Values
The `days_since_last_login` column contains 987 values greater than 21. According to the Data Dictionary, this value measures the number of days since the learner's last login during the first three weeks, measured at the end of week 3. Values greater than 21 therefore require investigation.

### 6. Video Watch Percentages Above 100%
38 learners have `video_watch_pct` values above 100%, with the maximum being 115.3%. The Data Dictionary explains that the video player can over-report this value when learners rewatch content. These values are therefore suspicious but may reflect known system behaviour rather than simple data-entry errors.

### 7. Inconsistent Education Categories
The `education` column contains multiple representations of apparently identical categories, such as `Bachelors`, `bachelors`, `Bachelor's`, and `bachelor degree`. Similar inconsistencies exist for high school, diploma, masters, and other categories. These labels will need to be standardized before modelling.

### 8. Inconsistent Mentor Values
The `has_mentor` column contains multiple representations of the same boolean values, including `0`, `No`, `FALSE`, `no`, `N`, `1`, `TRUE`, `Yes`, `Y`, and `yes`. These will need to be converted into a consistent representation.

### 9. Inconsistent Device Categories
The `device` column contains inconsistent labels and capitalization, including `mobile`, `Mobile`, `MOBILE`, `phone`, `laptop`, `Laptop`, `notebook`, `desktop`, `Desktop`, `PC`, `tablet`, `Tablet`, and `ipad`. These categories will need to be standardized.

### 10. Inconsistent City Tier Categories
The `city_tier` column contains numeric and text representations of the same categories. For example, Tier 2 appears as `2`, `T2`, `Tier 2`, and `tier-2`. Similar inconsistencies exist for Tiers 1 and 3. There are also 53 missing values.

### 11. Inconsistent Signup Date Formats
The `signup_date` column is stored as an object/text column and contains multiple date formats, including `YYYY-MM-DD`, `DD-MM-YYYY`, and `DD/MM/YYYY`. The column must be converted into a consistent datetime representation before reliable date-based analysis can be performed.

### 12. Future Information / Data Leakage
Two columns contain information that would not be available when the prediction is supposed to be made. `final_score` is calculated at the end of the 12-week program, and `certificate_issued` is recorded after the program ends. Since predictions are supposed to be made at the end of week 3, these columns must not be used as model inputs.

The `notes` column is also highly incomplete and unstructured, with approximately half of its values missing. It will likely be excluded from the initial model.


# Day 3 - Data Cleaning


In [37]:
clean_df = train_df.copy()

In [38]:
clean_df.shape

(2484, 19)

In [39]:
clean_df.head()

,learner_id,signup_date,age,city_tier,education,hours_per_week_pledged,weekly_logins_avg,assignments_submitted,forum_posts,video_watch_pct,has_mentor,device,referral_source,payment_plan,days_since_last_login,certificate_issued,final_score,notes,completed_course
0,L-01146,2026-01-22,24.0,Tier 3,OTHER,8.3,4.20,3,6,89.0,Y,MOBILE,instagram,scholarship,1,Y,78.6,NaN,1.0
1,L-00695,2026-06-02,15.0,Tier 1,Bachelors,1.0,2.54,0,4,31.2,no,desktop,instagram,paid,33,no,NaN,NaN,0.0
2,L-00530,20-05-2026,24.0,tier-2,Diploma,10.3,0.50,0,3,10.7,NaN,laptop,instagram,scholarship,23,No,NaN,follow up next intake,0.0
3,L-02228,2026-01-19,23.0,1,B.Tech,1.6,3.01,0,0,40.2,no,mobile,school_partner,free,33,FALSE,NaN,duplicate? check,0.0
4,L-00914,16/02/2026,21.0,2,High School,9.8,4.36,3,1,36.7,no,Desktop,instagram,paid,19,TRUE,67.1,follow up next intake,1.0


In [40]:
clean_df.columns.tolist()

['learner_id',
 'signup_date',
 'age',
 'city_tier',
 'education',
 'hours_per_week_pledged',
 'weekly_logins_avg',
 'assignments_submitted',
 'forum_posts',
 'video_watch_pct',
 'has_mentor',
 'device',
 'referral_source',
 'payment_plan',
 'days_since_last_login',
 'certificate_issued',
 'final_score',
 'notes',
 'completed_course']

In [41]:
clean_df.duplicated().sum()

np.int64(83)

In [42]:
clean_df = clean_df.drop_duplicates()

In [43]:
clean_df.shape

(2401, 19)

In [44]:
clean_df.duplicated().sum()

np.int64(0)

### Duplicate Records

The raw training dataset contained 83 exact duplicate rows. These rows were removed because they contained no additional information and would cause duplicated learners to receive extra weight during model training.

After removal, no exact duplicate rows remain.

In [45]:
clean_df[clean_df["learner_id"].duplicated(keep=False)]

,learner_id,signup_date,age,city_tier,education,hours_per_week_pledged,weekly_logins_avg,assignments_submitted,forum_posts,video_watch_pct,has_mentor,device,referral_source,payment_plan,days_since_last_login,certificate_issued,final_score,notes,completed_course
551,L-01467,27-05-2026,28.0,Tier 2,Bachelors,13.6,2.89,6,1,63.5,N,Mobile,ngo_outreach,free,3,TRUE,66.3,requested extension,1.0
939,L-01467,27-05-2026,28.0,Tier 2,Bachelors,13.6,2.89,6,1,63.5,N,Mobile,ngo_outreach,free,3,TRUE,66.3,requested extension,NaN


In [46]:
clean_df[clean_df["learner_id"] == "L-01467"]

,learner_id,signup_date,age,city_tier,education,hours_per_week_pledged,weekly_logins_avg,assignments_submitted,forum_posts,video_watch_pct,has_mentor,device,referral_source,payment_plan,days_since_last_login,certificate_issued,final_score,notes,completed_course
551,L-01467,27-05-2026,28.0,Tier 2,Bachelors,13.6,2.89,6,1,63.5,N,Mobile,ngo_outreach,free,3,TRUE,66.3,requested extension,1.0
939,L-01467,27-05-2026,28.0,Tier 2,Bachelors,13.6,2.89,6,1,63.5,N,Mobile,ngo_outreach,free,3,TRUE,66.3,requested extension,NaN


In [47]:
clean_df[clean_df["learner_id"] == "L-01467"].shape

(2, 19)

In [48]:
clean_df = clean_df[
    ~(
        (clean_df["learner_id"] == "L-01467") &
        (clean_df["completed_course"].isna())
    )
]

In [49]:
clean_df[clean_df["learner_id"] == "L-01467"]

,learner_id,signup_date,age,city_tier,education,hours_per_week_pledged,weekly_logins_avg,assignments_submitted,forum_posts,video_watch_pct,has_mentor,device,referral_source,payment_plan,days_since_last_login,certificate_issued,final_score,notes,completed_course
551,L-01467,27-05-2026,28.0,Tier 2,Bachelors,13.6,2.89,6,1,63.5,N,Mobile,ngo_outreach,free,3,TRUE,66.3,requested extension,1.0


In [50]:
clean_df["learner_id"].duplicated().sum()

np.int64(0)

### Duplicate Learner ID with Missing Outcome

After removing exact duplicate rows, learner `L-01467` still appeared twice. The two records had identical feature values, but one had `completed_course = 1` while the other had a missing completion outcome.

The duplicate record with the missing target was removed because an identical record containing the known outcome was already present.

In [51]:
clean_df["completed_course"].isna().sum()

np.int64(8)

In [52]:
clean_df = clean_df.dropna(subset=["completed_course"])

In [53]:
clean_df["completed_course"].isna().sum()

np.int64(0)

### Missing Target Values

The training data originally contained 9 rows with a missing `completed_course` value. These rows cannot be used for supervised learning because the correct outcome is unknown.

One of these rows was the duplicate `L-01467` record already removed during duplicate handling. The remaining 8 rows were removed from the cleaned training dataset.

In [54]:
clean_df.shape

(2392, 19)

In [55]:
clean_df["learner_id"].duplicated().sum()

np.int64(0)

In [56]:
clean_df["completed_course"].isna().sum()

np.int64(0)

In [57]:
clean_df[clean_df["age"] < 0][["learner_id", "age"]]

,learner_id,age
132,L-02530,-3.0
895,L-00503,-3.0
1090,L-01264,-3.0
1243,L-00261,-3.0
1469,L-02080,-3.0
1682,L-01445,-3.0


In [58]:
clean_df[clean_df["age"] > 100][["learner_id", "age"]]

,learner_id,age
30,L-02703,300.0
281,L-01185,199.0
481,L-02448,300.0
531,L-02461,300.0
622,L-00842,300.0
713,L-01071,199.0
841,L-01721,300.0
889,L-00090,300.0
902,L-01128,300.0
1254,L-01682,199.0


In [61]:
import numpy as np
clean_df.loc[
    (clean_df["age"] < 0) | (clean_df["age"] > 100),
    "age"
] = np.nan

In [62]:
(clean_df["age"] < 0) | (clean_df["age"] > 100)

,age
0,False
1,False
2,False
3,False
4,False
...,...
2478,False
2479,False
2480,False
2482,False


In [63]:
clean_df["age"].describe()

,age
count,2221.000000
mean,21.186853
std,4.121765
min,0.000000
25%,18.000000
50%,21.000000
75%,24.000000
max,36.000000


In [64]:
clean_df["age"].isna().sum()

np.int64(171)

### Invalid Age Values

The `age` column contained 22 clearly invalid values: 6 values of `-3` and 16 values of `199` or `300`.

Instead of removing the entire learner records, these invalid age values were replaced with missing values (`NaN`). This preserves the other useful information from those learners while allowing the invalid age values to be handled together with the existing missing ages later.

In [65]:
clean_df[clean_df["hours_per_week_pledged"] > 168][
    ["learner_id", "hours_per_week_pledged"]
]

,learner_id,hours_per_week_pledged
48,L-01420,318.3
178,L-02033,319.0
351,L-02360,233.8
663,L-00722,325.3
1351,L-00249,300.9
1441,L-02268,326.8
1553,L-02345,220.3
1624,L-01157,272.8


In [66]:
clean_df.loc[
    clean_df["hours_per_week_pledged"] > 168,
    "hours_per_week_pledged"
] = np.nan

In [67]:
clean_df["hours_per_week_pledged"].describe()

,hours_per_week_pledged
count,2278.000000
mean,7.405926
std,7.057229
min,1.000000
25%,4.800000
50%,7.100000
75%,9.500000
max,164.700000


In [68]:
clean_df["hours_per_week_pledged"].isna().sum()

np.int64(114)

### Invalid Weekly Hours

Nine learners had `hours_per_week_pledged` values greater than 168 hours per week, which is physically impossible because a week contains only 168 hours.

These values were treated as invalid and replaced with missing values rather than deleting the entire learner records.

In [69]:
clean_df["days_since_last_login"].describe()

,days_since_last_login
count,2392.000000
mean,19.208612
std,11.814099
min,0.000000
25%,10.000000
50%,18.000000
75%,27.000000
max,70.000000


In [70]:
clean_df["days_since_last_login"].value_counts().sort_index()

,count
days_since_last_login,
0,22
1,46
2,73
3,55
4,74
5,60
6,60
7,59
8,61


In [71]:
(clean_df["days_since_last_login"] > 21).sum()

np.int64(945)

In [72]:
clean_df.loc[
    clean_df["days_since_last_login"] > 21,
    "days_since_last_login"
] = np.nan

In [73]:
clean_df["days_since_last_login"].describe()

,days_since_last_login
count,1447.000000
mean,11.362820
std,6.148267
min,0.000000
25%,6.000000
50%,12.000000
75%,17.000000
max,21.000000


In [74]:
clean_df["days_since_last_login"].isna().sum()

np.int64(945)

In [75]:
clean_df["age"].describe()

,age
count,2221.000000
mean,21.186853
std,4.121765
min,0.000000
25%,18.000000
50%,21.000000
75%,24.000000
max,36.000000


In [76]:
clean_df["hours_per_week_pledged"].describe()

,hours_per_week_pledged
count,2278.000000
mean,7.405926
std,7.057229
min,1.000000
25%,4.800000
50%,7.100000
75%,9.500000
max,164.700000


In [77]:
clean_df["days_since_last_login"].describe()

,days_since_last_login
count,1447.000000
mean,11.362820
std,6.148267
min,0.000000
25%,6.000000
50%,12.000000
75%,17.000000
max,21.000000


In [78]:
clean_df["video_watch_pct"].describe()

,video_watch_pct
count,2249.000000
mean,46.749844
std,25.230107
min,0.000000
25%,28.800000
50%,45.500000
75%,63.900000
max,115.300000


In [79]:
clean_df[clean_df["age"] == 0][["learner_id", "age"]]

,learner_id,age
361,L-00454,0.0
368,L-00870,0.0
664,L-00919,0.0
814,L-02523,0.0
846,L-00616,0.0
950,L-01276,0.0
1642,L-02321,0.0


In [80]:
(clean_df["age"] == 0).sum()

np.int64(7)

In [81]:
clean_df.loc[
    clean_df["video_watch_pct"] > 100,
    "video_watch_pct"
] = 100

In [82]:
clean_df["video_watch_pct"].describe()

,video_watch_pct
count,2249.000000
mean,46.628457
std,24.945726
min,0.000000
25%,28.800000
50%,45.500000
75%,63.900000
max,100.000000


In [83]:
(clean_df["video_watch_pct"] > 100).sum()

np.int64(0)

In [90]:
clean_df.loc[
    (clean_df["age"] <= 0) | (clean_df["age"] > 100),
    "age"
] = np.nan

In [85]:
clean_df["age"].describe()

,age
count,2214.000000
mean,21.253839
std,3.952002
min,15.000000
25%,18.000000
50%,21.000000
75%,24.000000
max,36.000000


In [86]:
(clean_df["age"] <= 0).sum()

np.int64(0)

In [87]:
clean_df.loc[
    (clean_df["age"] <= 0) | (clean_df["age"] > 100),
    "age"
] = np.nan

In [89]:
(clean_df["age"] <= 0).sum()

np.int64(0)

In [91]:
clean_df["education"].value_counts(dropna=False)

,count
education,
Bachelors,605
High School,567
Diploma,378
Masters,166
Other,104
NaN,86
DIPLOMA,50
B.Tech,50
highschool,48


In [92]:
education_mapping = {
    "Bachelors": "Bachelors",
    "bachelors": "Bachelors",
    "Bachelor's": "Bachelors",
    "bachelor degree": "Bachelors",
    "B.Tech": "Bachelors",

    "High School": "High School",
    "highschool": "High School",
    "high school": "High School",
    "High-School": "High School",
    "HS": "High School",

    "Diploma": "Diploma",
    "diploma": "Diploma",
    "DIPLOMA": "Diploma",

    "Masters": "Masters",
    "masters": "Masters",
    "Master's": "Masters",
    "M.Sc": "Masters",

    "Other": "Other",
    "other": "Other",
    "OTHER": "Other",

    "unknown": np.nan
}

In [93]:
clean_df["education"] = clean_df["education"].replace(education_mapping)

In [94]:
clean_df["education"].value_counts(dropna=False)

,count
education,
Bachelors,769
High School,726
Diploma,466
Masters,208
Other,127
NaN,96


In [95]:
clean_df["education"].unique()

array(['Other', 'Bachelors', 'Diploma', 'High School', 'Masters', nan],
      dtype=object)

### Education Category Standardization

The `education` column contained inconsistent labels caused by different data-entry formats and system versions. Multiple labels represented the same education level.

The following categories were standardized:

- `Bachelors`, `bachelors`, `Bachelor's`, `bachelor degree`, and `B.Tech` → `Bachelors`
- `High School`, `highschool`, `high school`, `High-School`, and `HS` → `High School`
- `Diploma`, `diploma`, and `DIPLOMA` → `Diploma`
- `Masters`, `masters`, `Master's`, and `M.Sc` → `Masters`
- `Other`, `other`, and `OTHER` → `Other`
- `unknown` was converted to a missing value because it does not provide a meaningful education category.

This standardization prevents the model from treating different spellings or equivalent qualifications as separate categories.

In [96]:
clean_df["city_tier"].dtype

dtype('O')

In [99]:
city_tier_mapping = {
    1: 1,
    "1": 1,
    "T1": 1,
    "Tier 1": 1,
    "tier-1": 1,

    2: 2,
    "2": 2,
    "T2": 2,
    "Tier 2": 2,
    "tier-2": 2,

    3: 3,
    "3": 3,
    "T3": 3,
    "Tier 3": 3,
    "tier-3": 3
}

clean_df["city_tier"] = clean_df["city_tier"].replace(city_tier_mapping)

In [100]:
clean_df["city_tier"].value_counts(dropna=False)

,count
city_tier,
2.0,965
1.0,798
3.0,581
NaN,48


In [101]:
clean_df["city_tier"].unique()

array([ 3.,  1.,  2., nan])

In [102]:
clean_df["has_mentor"].value_counts(dropna=False)

,count
has_mentor,
0,297
No,290
FALSE,284
no,283
N,276
1,193
TRUE,186
Yes,180
Y,177


In [103]:
mentor_mapping = {
    0: 0,
    1: 1,

    "0": 0,
    "1": 1,

    "No": 0,
    "no": 0,
    "N": 0,
    "FALSE": 0,

    "Yes": 1,
    "yes": 1,
    "Y": 1,
    "TRUE": 1
}

In [106]:
clean_df["has_mentor"] = clean_df["has_mentor"].replace(mentor_mapping)

In [105]:
clean_df["has_mentor"].value_counts(dropna=False)

,count
has_mentor,
0.0,1430
1.0,900
NaN,62


In [107]:
clean_df["has_mentor"].unique()

array([ 1.,  0., nan])

In [108]:
clean_df["device"].value_counts(dropna=False)

,count
device,
mobile,986
laptop,575
desktop,223
tablet,116
Mobile,94
phone,85
MOBILE,80
notebook,71
Laptop,68


In [109]:
device_mapping = {
    "mobile": "Mobile",
    "Mobile": "Mobile",
    "MOBILE": "Mobile",
    "phone": "Mobile",

    "laptop": "Laptop",
    "Laptop": "Laptop",
    "notebook": "Laptop",

    "desktop": "Desktop",
    "Desktop": "Desktop",
    "PC": "Desktop",

    "tablet": "Tablet",
    "Tablet": "Tablet",
    "ipad": "Tablet"
}

In [110]:
clean_df["device"] = clean_df["device"].replace(device_mapping)

In [111]:
clean_df["device"].value_counts(dropna=False)

,count
device,
Mobile,1245
Laptop,714
Desktop,282
Tablet,151


In [112]:
clean_df["device"].unique()

array(['Mobile', 'Desktop', 'Laptop', 'Tablet'], dtype=object)

In [113]:
clean_df["signup_date"].dtype

dtype('O')

In [115]:
clean_df["signup_date"] = pd.to_datetime(
    clean_df["signup_date"],
    dayfirst=True,
    errors="coerce"
)

In [116]:
clean_df["signup_date"].head(20)

,signup_date
0,2026-01-22
1,2026-06-02
2,NaT
3,2026-01-19
4,NaT
5,2026-06-03
6,2026-04-09
7,NaT
8,2026-04-14
9,NaT


In [117]:
clean_df["signup_date"].dtype

dtype('<M8[ns]')

In [118]:
clean_df["signup_date"].isna().sum()

np.int64(739)

In [119]:
clean_df["signup_date"].min(), clean_df["signup_date"].max()

(Timestamp('2026-01-01 00:00:00'), Timestamp('2026-06-28 00:00:00'))

In [120]:
clean_df["signup_date"] = pd.to_datetime(
    train_df.loc[clean_df.index, "signup_date"],
    format="mixed",
    dayfirst=True,
    errors="coerce"
)

In [121]:
clean_df["signup_date"].isna().sum()

np.int64(0)

In [122]:
clean_df["signup_date"].head(20)

,signup_date
0,2026-01-22
1,2026-06-02
2,2026-05-20
3,2026-01-19
4,2026-02-16
5,2026-06-03
6,2026-04-09
7,2026-04-11
8,2026-04-14
9,2026-01-09


In [123]:
clean_df["signup_date"].min(), clean_df["signup_date"].max()

(Timestamp('2026-01-01 00:00:00'), Timestamp('2026-06-28 00:00:00'))

In [124]:
clean_df["weekly_logins_avg"].describe()

,weekly_logins_avg
count,2303.000000
mean,3.085089
std,1.498298
min,0.000000
25%,2.080000
50%,3.130000
75%,4.090000
max,7.000000


In [125]:
clean_df[
    (clean_df["weekly_logins_avg"] < 0) |
    (clean_df["weekly_logins_avg"] > 7)
][["learner_id", "weekly_logins_avg"]]

,learner_id,weekly_logins_avg


In [126]:
clean_df["weekly_logins_avg"].isna().sum()

np.int64(89)

In [127]:
clean_df["certificate_issued"].value_counts(dropna=False)

,count
certificate_issued,
no,304
No,304
0,298
FALSE,293
N,292
Y,193
TRUE,187
yes,181
Yes,175


In [128]:
clean_df["final_score"].describe()

,final_score
count,1128.000000
mean,70.343085
std,16.780981
min,13.300000
25%,62.100000
50%,74.250000
75%,82.025000
max,100.000000


In [129]:
clean_df["final_score"].isna().sum()

np.int64(1264)

In [130]:
clean_df["notes"].value_counts(dropna=False)

,count
notes,
NaN,1200
asked about certificate,224
follow up next intake,202
duplicate? check,200
-,196
requested extension,187
called on phone,183


In [131]:
clean_df.isna().sum().sort_values(ascending=False)

,0
final_score,1264
notes,1200
days_since_last_login,945
age,178
video_watch_pct,143
hours_per_week_pledged,114
education,96
weekly_logins_avg,89
has_mentor,62
city_tier,48


In [132]:
(clean_df.isna().mean() * 100).round(2).sort_values(ascending=False)

,0
final_score,52.84
notes,50.17
days_since_last_login,39.51
age,7.44
video_watch_pct,5.98
hours_per_week_pledged,4.77
education,4.01
weekly_logins_avg,3.72
has_mentor,2.59
city_tier,2.01


In [133]:
age_median = clean_df["age"].median()
age_median

21.0

In [134]:
clean_df["age"] = clean_df["age"].fillna(age_median)

In [135]:
clean_df["age"].isna().sum()

np.int64(0)

In [136]:
hours_median = clean_df["hours_per_week_pledged"].median()
hours_median

7.1

In [137]:
clean_df["hours_per_week_pledged"] = clean_df[
    "hours_per_week_pledged"
].fillna(hours_median)

In [138]:
clean_df["hours_per_week_pledged"].isna().sum()

np.int64(0)

In [139]:
logins_median = clean_df["weekly_logins_avg"].median()
logins_median

3.13

In [140]:
clean_df["weekly_logins_avg"] = clean_df[
    "weekly_logins_avg"
].fillna(logins_median)

In [141]:
clean_df["weekly_logins_avg"].isna().sum()

np.int64(0)

In [142]:
video_median = clean_df["video_watch_pct"].median()
video_median

45.5

In [143]:
clean_df["video_watch_pct"] = clean_df[
    "video_watch_pct"
].fillna(video_median)

In [144]:
clean_df["video_watch_pct"].isna().sum()

np.int64(0)

In [145]:
clean_df["days_since_last_login_missing"] = (
    clean_df["days_since_last_login"].isna().astype(int)
)

In [146]:
clean_df["days_since_last_login_missing"].value_counts()

,count
days_since_last_login_missing,
0,1447
1,945


In [147]:
login_days_median = clean_df["days_since_last_login"].median()
login_days_median

12.0

In [148]:
clean_df["days_since_last_login"] = clean_df[
    "days_since_last_login"
].fillna(login_days_median)

In [149]:
clean_df["days_since_last_login"].isna().sum()

np.int64(0)

In [150]:
clean_df["education"] = clean_df["education"].fillna("Unknown")

In [151]:
clean_df["education"].value_counts(dropna=False)

,count
education,
Bachelors,769
High School,726
Diploma,466
Masters,208
Other,127
Unknown,96


In [152]:
clean_df["city_tier"] = clean_df["city_tier"].astype("Int64").astype("string")

In [153]:
clean_df["city_tier"] = clean_df["city_tier"].fillna("Unknown")

In [154]:
clean_df["city_tier"].value_counts(dropna=False)

,count
city_tier,
2,965
1,798
3,581
Unknown,48


In [155]:
clean_df["has_mentor"] = clean_df["has_mentor"].fillna(-1)

In [156]:
clean_df["has_mentor"].value_counts(dropna=False)

,count
has_mentor,
0.0,1430
1.0,900
-1.0,62


In [159]:
age_median = clean_df["age"].median()
age_median

21.0

In [160]:
clean_df["age"] = clean_df["age"].fillna(age_median)

In [161]:
clean_df["age"].isna().sum()

np.int64(0)

In [162]:
hours_median = clean_df["hours_per_week_pledged"].median()
hours_median

7.1

In [165]:
clean_df["hours_per_week_pledged"] = clean_df[
    "hours_per_week_pledged"
].fillna(hours_median)

In [164]:
clean_df["hours_per_week_pledged"].isna().sum()

np.int64(0)

In [166]:
logins_median = clean_df["weekly_logins_avg"].median()
logins_median

3.13

In [167]:
clean_df["weekly_logins_avg"] = clean_df[
    "weekly_logins_avg"
].fillna(logins_median)

In [168]:
clean_df["weekly_logins_avg"].isna().sum()

np.int64(0)

In [169]:
video_median = clean_df["video_watch_pct"].median()
video_median

45.5

In [170]:
clean_df["video_watch_pct"] = clean_df[
    "video_watch_pct"
].fillna(video_median)

In [171]:
clean_df["video_watch_pct"].isna().sum()

np.int64(0)

In [172]:
clean_df["days_since_last_login_missing"] = (
    clean_df["days_since_last_login"].isna().astype(int)
)

In [173]:
clean_df["days_since_last_login_missing"].value_counts()

,count
days_since_last_login_missing,
0,2392


In [174]:
login_days_median = clean_df["days_since_last_login"].median()
login_days_median

12.0

In [175]:
clean_df["days_since_last_login"] = clean_df[
    "days_since_last_login"
].fillna(login_days_median)

In [176]:
clean_df["days_since_last_login"].isna().sum()

np.int64(0)

In [177]:
clean_df["days_since_last_login_missing"].value_counts()

,count
days_since_last_login_missing,
0,2392


In [178]:
clean_df["days_since_last_login_missing"] = (
    train_df.loc[clean_df.index, "days_since_last_login"].isna().astype(int)
)

In [179]:
clean_df["days_since_last_login_missing"].value_counts()

,count
days_since_last_login_missing,
0,2392


In [180]:
clean_df["days_since_last_login"].isna().sum()

np.int64(0)

In [181]:
raw_df = pd.read_csv("/content/cohort_train.csv")

In [182]:
raw_df["days_since_last_login"].isna().sum()

np.int64(0)

In [183]:
missing_login_ids = set(
    raw_df.loc[
        raw_df["days_since_last_login"].isna(),
        "learner_id"
    ]
)

In [184]:
clean_df["days_since_last_login_missing"] = (
    clean_df["learner_id"].isin(missing_login_ids).astype(int)
)

In [185]:
clean_df["days_since_last_login_missing"].value_counts()

,count
days_since_last_login_missing,
0,2392


In [186]:
print(raw_df.shape)
print(raw_df["days_since_last_login"].isna().sum())

(2484, 19)
0


In [187]:
raw_df["days_since_last_login"].value_counts(dropna=False).tail(10)

,count
days_since_last_login,
59,3
51,3
47,3
52,3
50,2
56,2
48,2
53,1
70,1


In [188]:
raw_df[
    raw_df["learner_id"].isin(
        ["L-00695", "L-00530", "L-02228", "L-02550"]
    )
][["learner_id", "days_since_last_login"]]

,learner_id,days_since_last_login
1,L-00695,33
2,L-00530,23
3,L-02228,33
8,L-02550,24


In [189]:
(raw_df["days_since_last_login"] > 21).sum()

np.int64(987)

In [190]:
(raw_df["days_since_last_login"] < 0).sum()

np.int64(0)

In [191]:
raw_df["days_since_last_login"].min(), raw_df["days_since_last_login"].max()

(0, 70)

In [192]:
invalid_login_ids = set(
    raw_df.loc[
        raw_df["days_since_last_login"] > 21,
        "learner_id"
    ]
)

In [193]:
clean_df["days_since_last_login_invalid"] = (
    clean_df["learner_id"].isin(invalid_login_ids).astype(int)
)

In [194]:
clean_df["days_since_last_login_invalid"].value_counts()

,count
days_since_last_login_invalid,
0,1447
1,945


In [195]:
invalid_login_ids = set(
    raw_df.loc[
        raw_df["days_since_last_login"] > 21,
        "learner_id"
    ]
)

In [196]:
clean_df["days_since_last_login_invalid"] = (
    clean_df["learner_id"].isin(invalid_login_ids).astype(int)
)

In [197]:
clean_df["days_since_last_login_invalid"].value_counts()

,count
days_since_last_login_invalid,
0,1447
1,945


In [199]:
clean_df["days_since_last_login"].isna().sum()

np.int64(0)

In [200]:
clean_df["days_since_last_login"].max()

21.0

### Days Since Last Login Validation

The `days_since_last_login` column contained no actual missing values in the original dataset. However, 987 values were greater than 21, while the Data Dictionary defines this field as the number of days since the learner's last login during the first three weeks of the program.

Because three weeks contains only 21 days, values greater than 21 were considered invalid. After duplicate records were removed, 945 invalid values remained.

These invalid values were replaced with the median valid value of 12 days. A separate `days_since_last_login_invalid` indicator was created to preserve information about whether the original value was invalid.

This prevents impossible values from being treated as genuine observations while retaining information about the underlying data-quality issue.

In [201]:
clean_df["education"] = clean_df["education"].fillna("Unknown")

In [202]:
clean_df["education"].isna().sum()

np.int64(0)

In [203]:
clean_df["education"].unique()

array(['Other', 'Bachelors', 'Diploma', 'High School', 'Masters',
       'Unknown'], dtype=object)

In [205]:
clean_df["city_tier"].value_counts(dropna=False)

,count
city_tier,
2,965
1,798
3,581
Unknown,48


In [206]:
clean_df["city_tier"].isna().sum()

np.int64(0)

In [207]:
clean_df["city_tier"] = clean_df["city_tier"].fillna("Unknown")

In [208]:
clean_df["city_tier"].isna().sum()

np.int64(0)

In [209]:
clean_df["has_mentor"] = clean_df["has_mentor"].fillna(-1)

In [210]:
clean_df["has_mentor"].isna().sum()

np.int64(0)

In [211]:
clean_df["has_mentor"].value_counts(dropna=False)

,count
has_mentor,
0.0,1430
1.0,900
-1.0,62


In [212]:
clean_df.isna().sum().sort_values(ascending=False)

,0
final_score,1264
notes,1200
age,0
signup_date,0
learner_id,0
education,0
city_tier,0
hours_per_week_pledged,0
weekly_logins_avg,0
video_watch_pct,0


In [213]:
clean_df = clean_df.drop(columns=["days_since_last_login_missing"])

In [214]:
clean_df.columns

Index(['learner_id', 'signup_date', 'age', 'city_tier', 'education',
       'hours_per_week_pledged', 'weekly_logins_avg', 'assignments_submitted',
       'forum_posts', 'video_watch_pct', 'has_mentor', 'device',
       'referral_source', 'payment_plan', 'days_since_last_login',
       'certificate_issued', 'final_score', 'notes', 'completed_course',
       'days_since_last_login_invalid'],
      dtype='object')

In [215]:
clean_df.shape

(2392, 20)

In [216]:
clean_df["learner_id"].duplicated().sum()

np.int64(0)

In [217]:
clean_df["age"].describe()

,age
count,2392.000000
mean,21.234950
std,3.802636
min,15.000000
25%,19.000000
50%,21.000000
75%,24.000000
max,36.000000


In [218]:
clean_df["hours_per_week_pledged"].describe()

,hours_per_week_pledged
count,2392.000000
mean,7.391346
std,6.887242
min,1.000000
25%,4.900000
50%,7.100000
75%,9.400000
max,164.700000


In [219]:
clean_df["weekly_logins_avg"].describe()

,weekly_logins_avg
count,2392.000000
mean,3.086760
std,1.470173
min,0.000000
25%,2.120000
50%,3.130000
75%,4.050000
max,7.000000


In [220]:
clean_df["video_watch_pct"].describe()

,video_watch_pct
count,2392.000000
mean,46.560995
std,24.189733
min,0.000000
25%,30.000000
50%,45.500000
75%,62.800000
max,100.000000


In [221]:
clean_df["days_since_last_login"].describe()

,days_since_last_login
count,2392.000000
mean,11.614548
std,4.791452
min,0.000000
25%,10.000000
50%,12.000000
75%,14.000000
max,21.000000


In [222]:
clean_df["completed_course"].value_counts()

,count
completed_course,
0.0,1501
1.0,891


In [223]:
leakage_columns = [
    "certificate_issued",
    "final_score",
    "notes"
]

leakage_columns

['certificate_issued', 'final_score', 'notes']

In [224]:
clean_df[
    clean_df["hours_per_week_pledged"] > 40
][["learner_id", "hours_per_week_pledged"]].sort_values(
    "hours_per_week_pledged",
    ascending=False
)

,learner_id,hours_per_week_pledged
2232,L-01231,164.7
1707,L-01052,146.4
2236,L-01127,140.7
2159,L-00911,138.8
1849,L-01111,99.1


In [225]:
(clean_df["hours_per_week_pledged"] > 40).sum()

np.int64(5)

In [226]:
clean_df["hours_per_week_pledged"].nlargest(20)

,hours_per_week_pledged
2232,164.7
1707,146.4
2236,140.7
2159,138.8
1849,99.1
2420,19.7
1979,18.5
2391,18.5
166,18.4
708,17.9


In [227]:
clean_df.loc[
    clean_df["hours_per_week_pledged"] > 40,
    "hours_per_week_pledged"
] = np.nan

In [228]:
clean_df["hours_per_week_pledged"].isna().sum()

np.int64(5)

In [229]:
hours_median = clean_df["hours_per_week_pledged"].median()
hours_median

7.1

In [230]:
clean_df["hours_per_week_pledged"] = clean_df[
    "hours_per_week_pledged"
].fillna(hours_median)

In [231]:
clean_df["hours_per_week_pledged"].isna().sum()

np.int64(0)

In [232]:
clean_df["hours_per_week_pledged"].max()

19.7

In [233]:
print("Shape:", clean_df.shape)
print("Duplicate learner IDs:", clean_df["learner_id"].duplicated().sum())
print("\nMissing values:")
print(clean_df.isna().sum().sort_values(ascending=False))

print("\nAge range:", clean_df["age"].min(), "to", clean_df["age"].max())
print("Hours/week range:",
      clean_df["hours_per_week_pledged"].min(),
      "to",
      clean_df["hours_per_week_pledged"].max())
print("Weekly logins range:",
      clean_df["weekly_logins_avg"].min(),
      "to",
      clean_df["weekly_logins_avg"].max())
print("Video watch range:",
      clean_df["video_watch_pct"].min(),
      "to",
      clean_df["video_watch_pct"].max())
print("Days since login range:",
      clean_df["days_since_last_login"].min(),
      "to",
      clean_df["days_since_last_login"].max())

print("\nTarget values:")
print(clean_df["completed_course"].value_counts())

Shape: (2392, 20)
Duplicate learner IDs: 0

Missing values:
final_score                      1264
notes                            1200
learner_id                          0
signup_date                         0
city_tier                           0
age                                 0
education                           0
hours_per_week_pledged              0
forum_posts                         0
video_watch_pct                     0
weekly_logins_avg                   0
assignments_submitted               0
device                              0
has_mentor                          0
referral_source                     0
payment_plan                        0
certificate_issued                  0
days_since_last_login               0
completed_course                    0
days_since_last_login_invalid       0
dtype: int64

Age range: 15.0 to 36.0
Hours/week range: 1.0 to 19.7
Weekly logins range: 0.0 to 7.0
Video watch range: 0.0 to 100.0
Days since login range: 0.0 to 21.0

Target valu

In [234]:
clean_df.to_csv("/content/cohort_train_cleaned.csv", index=False)

In [235]:
import os

os.path.getsize("/content/cohort_train_cleaned.csv")

260280

In [236]:
from google.colab import files

files.download("/content/cohort_train_cleaned.csv")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>